Simplify Eigen package config

ibab opened the pull request:
I think we can avoid specifying the Eigen commit hash in multiple places by using the `strip_prefix` option to `new_http_archive`.

Here, I've declared the real Eigen includes without a prefix, while the includes in `third_party` can only be reached by including `third_party/eigen3/...` (to avoid an infinite inclusion loop).
I've used this setup to successfully build tensorflow.

This can probably still be improved by not declaring the real Eigen includes at the top level.
Maybe it's possible to access them under `eigen_archive/...`, although I'm not sure how to set that up.

ibab committed:
Simplify Eigen package config

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented:
Sorry, there's a conflict I just introduced from another update to eigen libraries -- can you rebase and we can try this again?

ibab commented:
No problem, I've rebased the PR.

vrv commented on eigen.BUILD:
```diff
@@ -1,9 +1,8 @@
     name = "eigen",
-    hdrs = glob([archive_dir+"/**/*.h", archive_dir+"/unsupported/Eigen/*", archive_dir+"/unsupported/Eigen/CXX11/*", archive_dir+"/Eigen/*"]),
-    includes = [ archive_dir ],
+    hdrs = glob(["**/*.h", "unsupported/Eigen/CXX11/*", "Eigen/*"]),
```
are we missing an unsupported/Eigen/* ?

ibab replied:
Yeah, it seems like it was added in one of the last few commits, so it didn't make it into the PR before.

ibab committed:
Add missing unsupported/Eigen/*

vrv commented:
Sweet, let's give this a shot.  @tensorflow-jenkins test this please

vrv commented:
hm, that's an interesting failure.  Have you successfully been able to compile with this?

ibab commented:
I'm seeing the same problem now.
I'm pretty sure managed to compile it before the rebase, though :/

vrv commented:
The cpu test failed with a more appropriate error -- maybe we can resolve the CPU ones first and then figure out what's going on with GPU?

ibab commented:
I noticed that there's a new `third_party` Eigen header (at `third_party/eigen3/unsupported/Eigen/SpecialFunctions`) but adjusting it didn't help.

ibab commented:
Okay, the CPU tests are failing because I haven't adjusted the code that creates the pip package (It still assumes that eigen is located in `eigen_archive/eigen-eigen-\w+`).
Not sure about the GPU tests, though.

benoitsteiner commented on third_party/eigen3/unsupported/Eigen/CXX11/Tensor:
```diff
@@ -1 +1 @@
-#include "eigen-eigen-b4fa9622b809/unsupported/Eigen/CXX11/Tensor"
+#include "unsupported/Eigen/CXX11/Tensor"
```
You also need to adjust the new unsupported/Eigen/SpecialFunctions header

ibab committed:
Fix pip setup.py

ibab committed:
Adjust new eigen header

ibab commented:
I'm still not sure what's causing the problem with the GPU build.
Maybe the fact that Eigen is including the curand headers.
I'll have a closer look later.

ibab committed:
Fix bazel include dependency error

ibab committed:
Adjust Makefile to work with Eigen changes

ibab commented:
Finally got around to this.
It turns out that that the problem is caused by [a workaround in the GPU crosstool wrapper](https://github.com/tensorflow/tensorflow/blob/e95f4e760c6b6713b6b686ebeff9a1586a5831dd/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc#L103) that's used to support CUDA <= 6.0. Does TensorFlow still support those versions? If not, I could remove it.
I've added some extra code that makes the build work with the workaround.

Both problems from before should be solved now, so we can have another go at the tests :)

vrv commented:
I am fine dropping support for cuda 6.0, so go ahead!

ibab committed:
Remove nvcc workaround for CUDA <= 6.0

CUDA versions prior to 6.5 gave an

    error: kernel launches from templates are not allowed in system files

error when using gcc v4.8 and including code that uses templated
kernel launches via `-isystem`.
In order to work around this, the GPU crosstool converted `-isystem`
arguments containing the cuda headers into `-iquote` arguments.
This workaround has now been removed.

ibab commented:
Okay, I've removed that part of the crosstool script and pushed my changes.

vrv commented on tensorflow/contrib/makefile/Makefile:
```diff
@@ -38,8 +38,9 @@ HOST_OBJDIR := $(MAKEFILE_DIR)/gen/host_obj/
-# Find the current Eigen version name from the Bazel build file
-EIGEN_HASH := $(shell cat eigen.BUILD | grep archive_dir | head -1 | cut -f3 -d- | cut -f1 -d\")
+# Find the current Eigen version name from the cmake configuration
+EIGEN_HASH := $(shell cat tensorflow/contrib/cmake/external/eigen.cmake | \
```
I think it would still be nice to have the eigen hash in a core file rather than in contrib.

Since workspace.bzl is where the actual git repository is fetched from, what do you think about changing this script to look there?

ibab replied:
Yeah, that's a good idea.
I started doing that, but found it a bit brittle because the `workspace.bzl` doesn't state "this is the eigen hash" in a clear way.

I could define an `eigen_hash` variable in `workspace.bzl` and add a comment that this line is parsed from the makefile and cmake builds.

vrv commented:
Going to kick off a run just to see.  @tensorflow-jenkins test this please

vrv commented on tensorflow/workspace.bzl:
```diff
@@ -8,6 +8,7 @@ def tf_workspace(path_prefix = "", tf_repo_name = ""):
     name = "eigen_archive",
     url = "https://bitbucket.org/eigen/eigen/get/b4fa9622b809.tar.gz",
     sha256 = "2862840c2de9c0473a4ef20f8678949ae89ab25965352ee53329e63ba46cec62",
+    strip_prefix = "eigen-eigen-b4fa9622b809",
```
yeah, somehow separating the hash here would be great -- then we only have to replace the hash and the sha256 in one file and updating eigen becomes super easy :)

ibab committed:
Configure cmake and make to get eigen version from tensorflow/workspace.bzl

ibab commented:
I've pushed a new commit that isolates the `eigen_version` and `eigen_sha256` in the bazel config and adds some code to the cmake and make builds to extract those values.
I've used built-in tools in cmake to make sure that the config stays portable.

vrv commented:
Woohoo!  This is awesome.  

@tensorflow-jenkins test this please

cc @rmlarsen  -- our hacky scripts for updating eigen just got easier.

vrv commented:
Seems like an unrelated failure with cc_op_gen.cc, so merging.

ibab commented:
Thanks!
If there should be an issue with the cmake build as a result of this PR, you can post it here and I'll fix it.
